src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -84,130 +103,182 @@ private static SequencePosition GetBufferCrossSegment(int startIndex, in Sequenc
+        internal SequencePosition Seek(in SequencePosition start, in SequencePosition end, long count)
         {
             GetTypeAndIndices(start.GetInteger(), end.GetInteger(), out SequenceType type, out int startIndex, out int endIndex);
 
-            object startSegment = start.GetObject();
+            object startObject = start.GetObject();
             bool notInRange = endIndex - startIndex < count;
-            if (type == SequenceType.IMemoryList)
+            if (type == SequenceType.MultiSegment)
             {
-                object endSegment = end.GetObject();
-                if (notInRange || startSegment != endSegment)
+                object endObject = end.GetObject();
+                if (startObject != endObject)
                 {
-                    return SeekMultiSegment(startSegment, startIndex, endSegment, endIndex, count, checkEndReachable);
+                    Debug.Assert(startObject is ReadOnlySequenceSegment<T>);
+                    var startSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(startObject);
+
+                    int currentLength = startSegment.Memory.Length - startIndex;
+                    if (currentLength > count)
+                    {
+                        notInRange = false;
+                    }
+                    else
+                    {
+                        return SeekMultiSegment(startSegment, startIndex, endObject, endIndex, count - currentLength);
+                    }
                 }
             }
-            else if (notInRange)
+
+            if (notInRange)
                 ThrowHelper.ThrowArgumentOutOfRangeException_CountOutOfRange();
 
-            return new SequencePosition(startSegment, startIndex + (int)count);
+            return new SequencePosition(startObject, startIndex + (int)count);
         }
 
         [MethodImpl(MethodImplOptions.NoInlining)]
-        private static SequencePosition SeekMultiSegment(object startSegment, int startIndex, object endSegment, int endPosition, long count, bool checkEndReachable)
+        private static SequencePosition SeekMultiSegment(ReadOnlySequenceSegment<T> currentSegment, int startIndex, object endObject, int endPosition, long count)
         {
-            var current = (IMemoryList<T>)startSegment;
-            var end = (IMemoryList<T>)endSegment;
+            Debug.Assert(currentSegment != null);
+            Debug.Assert(currentSegment.Next != null);
 
-            SequencePosition result = default;
-            int currentIndex = startIndex;
-
-            do
+            if (count == 0)
             {
-                bool isCurrentAtEnd = current == end;
-
-                int memoryLength = isCurrentAtEnd ? endPosition - currentIndex : current.Memory.Length - currentIndex;
+                // End of segment. Move to start of next.
+                currentSegment = currentSegment.Next;
```

Comment:
If this ends up being null, won't this method fail?

Reply:
Can't ever be `null` (though did add `Debug.Assert(currentSegment.Next != null);` above)

Reply:
Although... it always does this in both if branches; can move it outside

Reply:
Because end cant be null?

Reply:
Yeah will throw if you create a ReadOnlySequence with a null end; so the only way it could get here is Slice with a manually provided end from on of these three methods
```
Slice(long start, SequencePosition end)
Slice(int start, SequencePosition end)
Slice(SequencePosition start, SequencePosition end)
```
However they do a bounds check ensuring `end` is before `_sequenceEnd` before calling into this method